src/include/columnar/columnar_storage.h

```diff
@@ -0,0 +1,54 @@
+/*-------------------------------------------------------------------------
+ *
+ * columnar_storage.h
+ *
+ * Type and function declarations for storage of columnar data in SMgr.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#ifndef COLUMNAR_STORAGE_H
+#define COLUMNAR_STORAGE_H
+
+#include "postgres.h"
+
+#include "utils/rel.h"
+
+/*
+ * Content of the first page in main fork, which stores metadata at file
+ * level.
+ */
+typedef struct ColumnarMetapage
+{
+	/*
+	 * Store version of file format used, so we can detect files from
+	 * previous versions if we change file format.
+	 */
+	int versionMajor;
+	int versionMinor;
+
+	/*
+	 * Each of the metadata table rows are identified by a storageId.
+	 * We store it also in the main fork so we can link metadata rows
+	 * with data files.
+	 */
+	uint64 storageId;
+
+	uint64 reserved_stripe_id;
+	uint64 reserved_row_number;
+	uint64 reserved_offset;
```

Comment:
1 ) I think we should handle upgrade scenario at least for `reserved_stripe_id` & `reserved_offset` (and need some tests in `multi_extension.sql` for such scenarios)

2 ) maybe `reserved_row_number` -> `reserved_stripe_row_number`

Reply:
I added some upgrade logic. It upgrades all columnar tables at extension update time, but if somehow that doesn't happen (e.g. if they don't update the extension and only install the new .so), it should catch the error and tell the user to vacuum (which will upgrade the metapage for that table only).